### ISSUE TITLE
State dump and verification

### ISSUE THREAD

**Issue by bowenwang1996** [nearprotocol/nearcore#2284]
Currently when we dump the state from a node, we always dump state as of the last block that the node has. @ailisp points out that this block might be on a fork that is not canonical. I think there are two related problems here:
* From which block should we dump state?
* How do we set the genesis height of the new network?

I think before we figure out upgradability, it is fine that when we do hard fork, the NEAR team *proposes* a new state from the current network that other validators can *verify*. In order to do this, I think we should dump the state as of the last final block. Since for a block to be final, it needs to go through pre-vote and pre-commit with 2/3 stake approval, it is safe to assume that this block is known by the majority of the validators and they can run the same state dump & migration script to reach the same state as proposed.

Since we would like to avoid potential double signing, I think it is best to not reusable a height for the new network that has appeared in the old network. However, when the state dump happens, there might still be blocks that are inflight in the network, which makes it hard to determine the latest height. I think for the genesis height of the new network, it might be reasonable to use latest  height from current node + some constant (like 10 for example) and it is fine if some heights are skipped. 

To summarize, I propose that we:
* dump state as of the last final block and publish the hash of the block from which we dump the state, together with the new state itself.
* set the genesis height of the new network to be the latest height from the node + some constant.

**Comment by mfornet** [comment]
We can use the same strategy used while syncing states. Usually to avoid forks we go back from the canonical chain a fixed number of blocks (50 in this case) and then find the first block in this epoch and use the state before applying the content of that block.

If we do it that way we always dump state just before starting some epoch (which can simplify complexity) and this state will be ready to be synced from nodes which are up to date.

**Comment by bowenwang1996** [comment]
@mfornet I think you are describing lightclient sync, which is something that we will implement in the future.

**Comment by mfornet** [comment]
AFAIU also full nodes will need to run sync state when they are validating new shards (on epoch change and seats redistribution). 

Anyway it makes sense to dump the state before applying first block from new epoch, since independently of block skips this state should be common to all nodes.

@bowenwang1996 also, how does this "+ constant" protect us, we still need to check that blocks belong to new chain, so other blocks that weren't build on top of new genesis should be discarded, isn't it?

**Comment by frol** [comment]
Can we implement some sort of soft shutdown? E.i., a day / a week before the restart we announce the epoch/block height that every node should stop processing, and thus the nodes automatically stop producing new blocks and this makes it safe to stop and dump without losing the transactions (those get rejected on submition).

**Comment by bowenwang1996** [comment]
@mfornet some external partners would prefer to have non-repeating height for the ease of auditing. Also it prevents people from potentially submit blocks from old network to slash people on the new network for double signing.

**Comment by bowenwang1996** [comment]
@frol I am ambivalent about whether the criterion should be time or height. The problem with choosing height as the trigger is that it is unknown how long it will take to reach that height, since the block production speed can vary. I think people generally want to have a clear idea of when the hard fork will happen and that's why I would prefer to shutdown nodes at a pre-announced time rather than height.

**Comment by ilblackdragon** [comment]
Blockchain time is blocks. It might vary, but it is only thing that everyone around the world can agree on. If you expect external people to do state dump / migration independently - then it needs to be on pre-agreed block height. And then we can just at this point shut down binary at that block height with expectation that hard fork is happening then.

Also for:
>  points out that this block might be on a fork that is not canonical.
We should be using latest final block from NFG.

**Comment by bowenwang1996** [comment]
> Blockchain time is blocks. It might vary, but it is only thing that everyone around the world can agree on. If you expect external people to do state dump / migration independently - then it needs to be on pre-agreed block height. And then we can just at this point shut down binary at that block height with expectation that hard fork is happening then.

If you read my proposal you will see that I don't expect everyone to run the state migration independently as that is really hard to coordinate (for example, to agree on the block from which the state should be dumped). To reiterate, I propose that the NEAR team dumps the state and publish it, along with the block hash of the block from which we dump the state, so that every validator should be able to verify that the state is correctly dumped. If we go with this approach I think it is better to stop the node at some pre-specified time so that people can know what to expect.

**Comment by bowenwang1996** [comment]
Okay it looks like people prefer to stop the node at a given height. So in practice it means that if we agree to stop at height `h`, the node should stop if there is a final block whose height is at least `h` and we dump the state from that block. It then means every validator can actually follow the same process and get to the same state dump, which makes things even easier for them.

We should implement this for testnet. More specifically, assuming that we have a one-week release cycle from now until mainnet, we need to release every `604800` blocks (assume 1b/s). This means that we need to:
* Have a service that watches the node through rpc and stop the node if it reaches the criterion mentioned above.
* Publish the next stop height somewhere public. Ideally we should generate automated notification it we are, say, within 10000 blocks of the stop height.

**Comment by frol** [comment]
Should we somehow bake this graceful node auto-shutdown mechanics into the node implementation? This way the upgrade can be automated with a simple script like:

```bash
while true; do
  nearup upgrade
  nearup start

  sleep 10
done
```

This will work if `nearup start` exists with the node which willingly shut itself down on hard fork condition.

**Comment by bowenwang1996** [comment]
I would prefer not to do that. We should only have code in nearcore that is necessary for the node to function.

**Comment by frol** [comment]
Well, otherwise if an external service misses the right time and doesn't kill the node, we may lose some transactions in the forks or may even have them finalized if the service acts unreliably across the network.

I would consider this feature to be the part of the nearcore upgradability protocol, and thus belong to the node implementation. :man_shrugging:

**Comment by bowenwang1996** [comment]
You don't have to shut it down at the exact right time. Since this height is announced ahead of time, any transaction on the chain past the announced height will not be considered, so I don't think there is a problem there.

**Comment by frol** [comment]
@bowenwang1996 It is unclear how the node makes a decision not to consider the transactions beyond the announced block height. Does this mean that it will stop block production, but still serve read-only RPC requests and internal routing?

**Comment by bowenwang1996** [comment]
@frol The node doesn't have to do it. Since the announcement of the height is public, people should be aware that once the height is close they should stop sending important transactions. Of course this isn't the long term solution and it will be much more smooth once we figure out upgradability.

**Comment by ailisp** [comment]
I think @frol suggested auto upgrade should be necessary for most users. For us and for advanced users it's reasonable to have migration and verification logic all in nearcore as Bowen said.

So
```
upgrade = clear data & download new genesis
migration = state_dump(), run migration scripts to create new genesis, upload genesis
verification = state_dump(up_to_block_hash), run migration scripts to create new genesis, download official genesis and compare.
```
are three different scripts

**Comment by frol** [comment]
@ailisp that is part of what I was suggesting. Another part was to make the upgrade process loss-free by designing a soft-stop feature preventing acknowledging the transactions after the block height chosen for upgrade.

**Comment by nearmax** [comment]
> Currently when we dump the state from a node, we always dump state as of the last block that the node has.

Are we currently dumping state the old way when we would just save the entire state column? Or the new way where we would convert the state into the state records? The old way includes state for all blocks that were not garbage collected yet. Any state root from the state dump then can be used for the genesis block.

> Since we would like to avoid potential double signing, I think it is best to not reusable a height for the new network that has appeared in the old network.

Reusing height creates double signing only if we allow supermajority to approve both forks. Instead we can explicitly prohibit nodes to finalize fork for some time which would give us a window to switch the forks without causing double-signing.

Also, we cannot reset the height since contracts can be recording the block height internally or relying on it to be strictly increasing. 

> some constant (like 10 for example) and it is fine if some heights are skipped.

Can our chain code deal with skipped heights?

> However, when the state dump happens, there might still be blocks that are inflight in the network, which makes it hard to determine the latest height. I think for the genesis height of the new network, it might be reasonable to use latest height from current node + some constant (like 10 for example) and it is fine if some heights are skipped.

Not only that, we might have blocks in flight that are already finalized but not included in our state dump.

---

I see it as a fundamental trade-off between two evils. In order to be able to resume the new code from some old state we need to do either of the following:
* Temporarily suspend finalization while we are proposing the fork from some finalized old state (this needs to be done to avoid double-signing);
* Revert the state of the finalized blocks losing some progress that the contracts did (e.g. let's say we are proposing state from block X, but block X+Y is finalized while we are doing it. Even if we resume from X+Y+Z height using state from block X, we are going to lose finalized state that happened in block X+Y).

Which one do we choose?

**Comment by nearmax** [comment]
@bowenwang1996 Are you currently suggesting that we are going to have a time interval during which no transactions will be considered, while we will be trying to get an agreement from all validators to hard fork from certain state? Sounds like it will take at least hours, because it will be social consensus. Are we okay with our network stopping accepting transactions for several hours at least?

**Comment by bowenwang1996** [comment]
@nearmax the plan is to stop considering transactions after a given height. I think the consensus should be reached beforehand (since we will follow a regular schedule it is very predictable), and, for those who didn't agree to do so, they naturally fork off by not stopping their node. If you are talking about the case where there is a critical bug and it takes time to reach a social consensus on what to do, then yes it will likely take several hours, but I think that's inevitable.

**Comment by ilblackdragon** [comment]
@bowenwang1996 when you say "communicated ahead of time" and "agreed on stopping" - that needs to be somehow implemented in the code as well.
E.g. nodes should indeed stop accepting blocks after that time appeared, otherwise how does a RPC user (a UI for cryptocorgi deployed to IPFS) becomes aware of it?

I like the idea of soft shutdown, @SkidanovAlex have suggested it before as well.
I think only block number is relevant in this case, yes it's not very predictable from wall clock perspective - but we don't want validators to be running this scripts manually anyway.
 We should optimize for this to be done automatically, and:
 - communicate `block_height` that should stop at in some smart contract or specified ahead of time for regular releases. + 
 - validators ahead of time review the code that was released into stable, verify that testnet runs and doesn't steal money / fails.
 - download and prepare their setup with new code
 - at that `height` their nodes shut down
 - all validators know the latest state root at that height, can dump their state, run migrations and restar the node.

This also fits well with future upgradability - https://commonwealth.im/near/proposal/discussion/261-upgradability?comment=649 having the same (a) and (b) and we will just switch (c) from stop everyone & migrate & restart everyone to stop independently & migrate & restart independently.

**Comment by bowenwang1996** [comment]
@ilblackdragon yes the question is whether to implement the logic of stopping at some height in nearcore. It feels to me that we should not implement unnecessary logic in nearcore, but rather have a separate service that watches the node and shut it down when it reaches a given height.

**Comment by bowenwang1996** [comment]
@ilblackdragon one issue that just occurred to me: if we decide to stop the node and dump state on a certain height, how should we set the new genesis time?

**Comment by ailisp** [comment]
Seems new genesis time should be new. I propose: assume we scheduled fork
happens at block height X, we should also scheduled a "new network start
time" and set new genesis time to that

**Comment by bowenwang1996** [comment]
@ailisp we cannot know at the current release the time for the next release because block time is not constant. Also let's not close issues that are not actually resolved.

**Comment by bowenwang1996** [comment]
@ailisp yes my concern is that we should not the gap to be too low. Otherwise people might not have finished their state dump process. But we cannot also set it to be too high as we don't want to have the network down for an extended period of time. So it may be difficult to set the new genesis time.

**Comment by bowenwang1996** [comment]
I propose that the new genesis time should be the block time of the block from which we dump the state + some constant (like 1 hour). If we go with this approach, we need to figure out what the constant should be and check with validators on whether it works for them. cc @jimmy3dita @ilblackdragon

**Comment by ilblackdragon** [comment]
Yes, just let's try to get to waaay smaller delta from last block's timestamp.
Ideally it should be defined by our tooling - how long does it take for us to dump / restart?

**Comment by bowenwang1996** [comment]
@ilblackdragon I think it was about 5 minutes.

**Comment by ailisp** [comment]
Does it take longer in a bigger state in the testnet?

**Comment by bowenwang1996** [comment]
@ailisp 5 minutes is the time it took for last testnet state migration, at which point it already has more than 100MB state. But yes, we should experiment with it and see if 5 minutes is reasonable.

**Comment by bowenwang1996** [comment]
Created issue #2447 for the case when there are more than one shards.

**Comment by ilblackdragon** [comment]
At this point, is not blocking Phase 0 launch, moving the rest to Phase 1.

**Comment by bowenwang1996** [comment]
Looks like we are going with the upgradability approach #2676 so we no longer need this. Instead, we should work on #2498.
